app/src/test/java/org/mozilla/fenix/ext/StringTest.kt

```diff
@@ -41,4 +52,209 @@ class StringTest {
+
+        assertThat("https://examplecorp.com@attacker.example/".shortened()).doesNotContain("examplecorp")
+        assertThat("https://examplecorp.com@attacker.example/".shortened()).doesNotContain("com")
+        assertThat("https://user:password@example.com/".shortened()).doesNotContain("user")
+        assertThat("https://user:password@example.com/".shortened()).doesNotContain("password")
+    }
+
+    @Test
+    fun `eTLDs should not be dropped`() {
+        // See https://bugzilla.mozilla.org/show_bug.cgi?id=1554984#c11
+        "http://mozfreddyb.github.io/" shortenedShouldBecome "mozfreddyb.github.io"
+        "http://web.security.plumbing/" shortenedShouldBecome "web.security.plumbing"
+    }
+
+    @Test
+    fun `ipv4 addresses should be returned as is`() {
+        // See https://bugzilla.mozilla.org/show_bug.cgi?id=1554984#c11
+        "192.168.85.1" shortenedShouldBecome "192.168.85.1"
+    }
+
+    @Test
+    fun `about buildconfig should not be modified`() {
+        // See https://bugzilla.mozilla.org/show_bug.cgi?id=1554984#c11
+        "about:buildconfig" shortenedShouldBecome "about:buildconfig"
+    }
+
+    @Test
+    fun `encoded userinfo should still be considered userinfo`() {
+        "https://user:password%40really.evil.domain%2F@mail.google.com" shortenedShouldBecome "mail.google.com"
+    }
+
+    @Test
+    @Ignore("This would be more correct, but does not appear to be an attack vector")
+    fun `should decode DWORD IP addresses`() {
+        "https://16843009" shortenedShouldBecome "1.1.1.1"
+    }
+
+    @Test
+    @Ignore("This would be more correct, but does not appear to be an attack vector")
+    fun `should decode octal IP addresses`() {
+        "https://000010.000010.000010.000010" shortenedShouldBecome "8.8.8.8"
+    }
+
+    @Test
+    @Ignore("This would be more correct, but does not appear to be an attack vector")
+    fun `should decode hex IP addresses`() {
+        "http://0x01010101" shortenedShouldBecome "1.1.1.1"
+    }
+
+    // BEGIN test cases borrowed from desktop (shortUrl is used for Top Sites on new tab)
+    // Test cases are modified, as we show the eTLD
+    // (https://searchfox.org/mozilla-central/source/browser/components/newtab/test/unit/lib/ShortUrl.test.js)
+    @Test
+    fun `should return a blank string if url is blank`() {
+        "" shortenedShouldBecome ""
+    }
+
+    @Test
+    fun `should return the 'url' if not a valid url`() {
+        "something" shortenedShouldBecome "something"
+        "http:" shortenedShouldBecome "http:"
+        "http::double-colon" shortenedShouldBecome "http::double-colon"
+        // The largest allowed port is 65,535
+        "http://badport:65536/" shortenedShouldBecome "http://badport:65536/"
+    }
+
+    @Test
+    fun `should convert host to idn when calling shortURL`() {
+        "http://$PUNYCODE.blah.com" shortenedShouldBecome "$IDN.blah.com"
```

Comment:
The Punycode to IDN rules seem quite complicated. I think we're supposed to check that all of the IDN uses the same codepage or else show Punycode instead.

Reply:
They do seem pretty complex.  Do you mind if I leave a comment describing that change, but push implemention to a TODO?  IMO, it should go through prioritization and sizing.

Reply:
Spoke with colintheshots about this and opened [a follow up issue](https://github.com/mozilla-mobile/fenix/issues/6642) for this change.